activerecord/lib/active_record/connection_adapters/abstract/transaction.rb

```diff
@@ -498,6 +498,11 @@ def within_new_transaction(isolation: nil, joinable: true)
             raise
           ensure
             unless error
+              # In 7.1 we enforce timeout >= 0.4.0 which now longer use throw, so we can
+              # go back to the original behavior of committing on non-local return.
+              # If users are using throw, we assume it's not an error case.
+              completed = true unless ActiveRecord.rollback_transaction_on_non_local_return
+
               if Thread.current.status == "aborting"
                 rollback_transaction
               elsif !completed && transaction.written
```

Comment:
We're adding a config option so eager applications can skip the cycle and get straight back to the better behaviour, but I think we should still immediately deprecate the rollback behaviour -- as long as it's configurable, no library etc can assume either behaviour.

i.e. I think this branch should be warning before it rolls back. (And the one below should still-silently switch back to commit.)